_posts/2016-05-11-android-testing-with-kotlin.md

```diff
@@ -0,0 +1,422 @@
+just as it is with variable declaration. But as Kotlin has first class support for
+user defined **infix** functions, as seen above, we can enrich the standard library
+with syntactic sugar of our choosing with no effort whatsoever. For example here we
+have replaced
+
+```java
+assertEquals(expected, actual)
+```
+
+checks with
+
+```java
+actual equals expected
+```
+
+statements for readability, utilizing our custom infix procedure `equals`.
+
+```java
+infix fun Any.equals(expected: Any) {
+    assertEquals(expected, this)
+}
+```
+
+Of course, all other gems of Kotlin are readily available to us, including
+convenient array declaration, avoidance of clumsy `for` loops using `map` and
+other functional elements.
+
+Last but not least, we annotate out test functions with `@Test` as usual to make
+compiler aware of them.
+
+```java
+@Test
+fun myTestCase() { ... }
+```
+
+## Espresso tests
+
+Making use of the interoperability of Java and Kotlin code we can blend Espresso into
+our Kotlin test classes in the blink of an eye. That's how we can benefit both from
+battle tested Espresso framework that makes testing UI components a bliss and
+all the gems that Kotlin provides as a language to make writing our _glue_ code and other
+testing logic as easy and clean as possible.
+
+As usual, we store our Espresso and other UI tests under `src/androidTest`, with a
+little discrepancy that instead of `java` subdirectory, we store our Kotlin source code in
+`kotlin` subdirectory. And in order to run Espresso tests we need to declare it as a dependency
+in our build script plus set `AndroidJUnitRunner` as the default test instrumentation
+runner, exactly like you would do with plain Espresso tests:
+
+```groovy
+android {
+    ...
+    defaultConfig {
+        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
+    }
+}
+
+dependencies {
+    ...
+    androidTestCompile 'com.android.support.test.espresso:espresso-core:2.2.2'
+}
+```
+
+Now to the test suite itself. Our example Espresso test class
+[`EspressoKotlinTest`](https://github.com/GreenhouseCI/kotlin-tests/blob/master/app/src/androidTest/kotlin/com/greenhouseci/kotlin_tests/kotlintests/EspressoKotlinTest.kt)
+looks like
+
+```ruby
+@RunWith(AndroidJUnit4::class)
+class EspressoKotlinTest {
+
+    ...
+
+    @Rule @JvmField
+    var activityRule = ActivityTestRule<MainActivity>(MainActivity::class.java)
+
+    ...
+
+    @Test
+    fun testAddItemsToList() {
+        val formatter = Formatter()
+        val lv = onData(anything()).inAdapterView(withId(R.id.lv_items))
+
+        var items = arrayOf(
+            Pair("    item 1", "item 1"),
+            Pair("item 2  ", "item 2"),
+            Pair("  item  \n3  ", "item 3")
+        )
+
+        items.map { pair -> addItemToList(pair.first) }
+        items.mapIndexed { index, pair ->
+            val expected = formatter.strip(pair.second)
```

Comment:
I'm probably misunderstanding this, but why is `pair.second` passed through the formatter? Shouldn't it be hard-coded as the expected value? Otherwise, if `pair.first` is sent to the formatter as part of being added to the list, and `pair.second` as part of this test, then the test would pass even with a broken formatter